R/read_alevin.R

```diff
@@ -115,29 +106,101 @@ read_usa_mode <- function(quant_dir,
+    stop(paste0("Missing Alevin output file(s): ", missing_files))
+  }
+
+  txi <- suppressMessages(tximport::tximport(
+    file.path(quant_dir, "alevin", "quants_mat.gz"),
+    type = "alevin"
+  ))
+  counts <- txi$counts
+}
+
+#' Read alevin metadata from json files
+#'
+#' @param quant_dir Path alevin output directory.
+#'
+#' @return A list containing alevin run metadata,
+#'   with NULL values for missing elements.
+#'
+#' @noRd
+read_alevin_metadata <- function(quant_dir){
+  cmd_info_path <- file.path(quant_dir, "cmd_info.json")
+  permit_json_path <- file.path(quant_dir, "generate_permit_list.json")
+  collate_json_path <- file.path(quant_dir, "collate.json")
   quant_json_path <- file.path(quant_dir, "quant.json")
   if(!file.exists(quant_json_path)){
     # file for alevin-fry < 0.4.1
     quant_json_path <- file.path(quant_dir, "meta_info.json")
-    if(!file.exists(quant_json_path)){
-      stop("Missing quant.json (or meta_info.json) in Alevin output directory")
-    }
   }
 
-  # check that USA mode is true in JSON file
-  quant_json <- jsonlite::fromJSON(quant_json_path)
-  if(quant_json$usa_mode != "TRUE"){
-    stop("Output files not in USA mode")
+  # get cmd_info, which should always be present
+  if (file.exists(cmd_info_path)){
+    cmd_info <- jsonlite::read_json(cmd_info_path)
+  } else {
+    stop("cmd_info.json is missing")
   }
 
-  # read in .mtx files
-  mtx <- Matrix::readMM(file = file.path(quant_dir, "alevin", "quants_mat.mtx"))%>%
-    Matrix::t() %>%
-    as("dgCMatrix")
-  cols <- readLines(file.path(quant_dir, "alevin", "quants_mat_cols.txt"))
-  rows <- readLines(file.path(quant_dir, "alevin", "quants_mat_rows.txt"))
-  dimnames(mtx) <- list(cols, rows)
+  # Read other info files if they exist. Otherwise, create dummy values
+  if (file.exists(permit_json_path)){
+    permit_info <- jsonlite::read_json(permit_json_path)
+  } else {
+    permit_info <- list()
+  }
+  if (file.exists(quant_json_path)){
+    collate_info <- jsonlite::read_json(collate_json_path)
+  } else {
+    collate_info <- list()
+  }
+  if (file.exists(quant_json_path)){
+    quant_info <- jsonlite::read_json(quant_json_path)
+  } else {
+    quant_info <- list()
+  }
 
-  counts <- collapse_intron_counts(mtx, which_counts)
-  return(counts)
+  # Create a metadata list
+  meta <- list(salmon_version = cmd_info[['salmon_version']],
+               reference_index = cmd_info[['index']])
+
+  # if we have permit_info data, we used alevin-fry, otherwise alevin
+  if (length(permit_info) == 0){
+    meta$mapping_tool <- "alevin"
+  } else {
+    meta$mapping_tool <- "alevin-fry"
+  }
+
+  # Add other metadata
+  # assume all alevin-fry tool versions are the same
+  meta$alevinfry_version <- permit_info[['version_str']]
+  meta$af_permit_type <- permit_info[['permit-list-type']]
+  meta$af_resolution <- quant_info[['resolution_strategy']]
+  meta$usa_mode <- quant_info[['usa_mode']]
```

Comment:
```suggestion
  meta$usa_mode <- quant_info[['usa_mode']]
  meta$transcript_type <- which_counts
```
If you like my earlier suggestion of including the transcript type in this function.